gusto/recovery.py

```diff
@@ -0,0 +1,318 @@
+        if boundary_method is not None:
+            if boundary_method != 'density' and boundary_method != 'velocity':
+                raise ValueError("Specified boundary_method % not valid" % boundary_method)
+            if VDG is None:
+                raise ValueError("If boundary_method is specified, VDG also needs specifying.")
+
+            # now specify things that we'll need if we are doing boundary recovery
+            if boundary_method == 'density':
+                # check dimensions
+                if self.V.value_size != 1:
+                    raise ValueError('This method only works for scalar functions.')
+                self.boundary_recoverer = Boundary_Recoverer(self.v_in, self.v_out, self.v)
+            if boundary_method == 'velocity':
+                # check dimensions
+                if self.V.value_size != 2:
+                    raise ValueError('This method only works for 2D vector functions.')
+                # declare spaces and functions manually for the scalar field
+                self.VDG0 = FunctionSpace(self.VDG.mesh(), "DG", 0)
+                self.VCG1 = FunctionSpace(self.VDG.mesh(), "CG", 1)
+                self.VDG1 = FunctionSpace(self.VDG.mesh(), "DG", 1)
+                self.v_in_scalar = Function(self.VDG0)
+                self.v_scalar = Function(self.VDG1)
+                self.v_out_scalar = Function(self.VCG1)
+
+                self.boundary_recoverer = Boundary_Recoverer(self.v_in_scalar, self.v_out_scalar, self.v_scalar)
+                # the boundary recoverer needs to be done on a scalar fields
+                # so need to extract component and restore it after the boundary recovery is done
+                self.project_to_vector = Projector(as_vector([self.v_out_scalar, self.v_out[1]]), self.v_out)
+                self.project_to_scalar_DG = Projector(self.v_in[0], self.v_in_scalar)
+                self.project_to_scalar_CG = Projector(self.v_out[0], self.v_out_scalar)
+                # need an extra averager that works on the scalar fields rather than the vector one
+                self.extra_averager = Averager(self.v_scalar, self.v_out_scalar)
+
+    def extract_scalar(self):
+        self.project_to_scalar_DG.project()
+        self.project_to_scalar_CG.project()
+
+    def restore_vector(self):
+        self.project_to_vector.project()
+
+    def project(self):
+        """
+        Perform the fully specified recovery.
+        """
+
+        if self.interpolator is not None:
+            self.interpolator.interpolate()
+        self.averager.project()
+        if self.boundary_method is not None:
+            if self.boundary_method == 'velocity':
+                self.extract_scalar()
+                self.boundary_recoverer.apply()
+                self.extra_averager.project()
+                self.restore_vector()
+            elif self.boundary_method == 'density':
+                self.boundary_recoverer.apply()
+                self.averager.project()
```

Comment:
What happens if I set `self.boundary_method = "cinnamon_buns"`? Perhaps there should be an `else` clause which rules out methods that aren't acceptable. In which case, that would be a `ValueError`.